### ISSUE TITLE
Training on zero terminal SNR seems to be broken

### ISSUE THREAD

**Issue by bghira** [huggingface/diffusers#4636]
### Describe the bug

I've used the Diffusers pix2pix script, adapted for training the SDXL U-net with a focus on text2img to do experiments over the last week.

I've observed a continual contrast issue with SDXL on v_prediction.

Prompt: `white woman in white pants and white shirt standing in front of white background`
![image](https://github.com/huggingface/diffusers/assets/59658056/b7c71656-df75-4be3-a9f1-fb616488fd9e)

# Versus SD 2.0-v (not 2.1) 

**After fine-tuning**
![image](https://github.com/huggingface/diffusers/assets/59658056/868fdff1-83bf-41e1-9e52-41fce52d8e6c)

**SD 2.0-v trained on portrait aspect**
![image](https://github.com/huggingface/diffusers/assets/59658056/56dadb35-889d-4cc2-a035-073e5418dde6)

# Versus SD 2.1

**(before terminal SNR fine-tuning)**
![image](https://github.com/huggingface/diffusers/assets/59658056/0dbc5d9d-7aa8-48f5-915d-0ade2c4f76e2)

**during terminal SNR fine-tuning:**
![image](https://github.com/huggingface/diffusers/assets/59658056/9a7486e5-52fc-41df-82dc-400a7373629a)

**after fine-tuning (sorry for different resolutions, that doesn't impact results):**
![image](https://github.com/huggingface/diffusers/assets/59658056/2116a7ad-5487-4fab-8118-442de470aa6c)

# Darkness prompt

**SD 2.0-v during fine-tuning, mild contrast issue**
![image](https://github.com/huggingface/diffusers/assets/59658056/270200b4-126a-4308-8322-26288e4f9892)

**SD 2.1-v after fine-tuning, perfect contrast**
![image](https://github.com/huggingface/diffusers/assets/59658056/d0be941c-abac-4cd5-b668-01d52eb8e7b4)

**SDXL 1.0**
![image](https://github.com/huggingface/diffusers/assets/59658056/b0862ace-9b88-4faa-961c-b7bea2d5893f)

### Reproduction

I've tried pretty much every tunable flag and code modification that made sense, and even some that didn't.

I've connected to other trainers who are familiar with terminal SNR training, that are just as lost and confused as I am.

This occurs:

* When training SDXL on `prediction_type='v_prediction'`. I have not tried `sample` (x-prediction).
* With or without EMA
* With any value of guidance_rescaling from 0.0 to 1.0
* With any value of guidance scaling, from 1.0 to 20.0
* Over many training days (1 week) or a quick-burn with a high learning rate
* LR scheduling has zero effect on this problem
* With low or high batch size and zero or more gradient accumulation steps
* Min SNR weighting doesn't have any impact on contrast, though it encouraged convergence
* Offset noise causes more stability issues. It's _possibly_ adding noise on a zero-sigma timestep?
* Input perturbations don't seem to help at all. 
* When using `timestep_spacing='leading'` or `timestep_spacing='trailing'`. I have not tried `linspace`.
* When using DDPMScheduler for training with `trained_betas` originating from DDIMScheduler with rescale parameter set.
* At any training resolution.
* When using AdamW8Bit or D-Adaptation (external dependency)

I have not been able to try disabling gradient checkpointing, which is theorised in some communities to have an impact on v_prediction adaptation - if it did, that would be a bug. I haven't been able to confirm.

I have heard anecdotal evidence that other trainers (eg. Khoya) have the same issue with SDXL v-prediction.

It's possible that this will simply never work, but that doesn't feel correct.

### System Info

- `diffusers` version: 0.20.0.dev0
- Platform: Linux-5.19.0-45-generic-x86_64-with-glibc2.31
- Python version: 3.9.16
- PyTorch version (GPU?): 2.1.0.dev20230730+cu118 (True)
- Huggingface_hub version: 0.16.4
- Transformers version: 4.30.2
- Accelerate version: 0.18.0
- xFormers version: 0.0.21+82254f4.d20230731
- Using GPU in script?: A100-80G
- Using distributed or parallel set-up in script?: No

### Who can help?

@williamberman and @PeterL1n

**Comment by bghira** [comment]
It took about 2600 steps of training on SD 2.1 for the new noise schedule to be adapted and in more than 3600 steps of training SDXL at the same resolution/batchsize and dataset contents, it just doesn't want to work.

evidently, SD 2.0 and 1.5 convert to v-prediction just fine.

**Comment by bghira** [comment]
@patrickvonplaten i've had SargeZT looking over the implementation details of the rescaled CFG code and the one thing that stood out was that the std_cfg is derived differently from the original code, that passes it in directly to use for calculating the rescaled guidance.

no conclusions yet, but maybe it's related to the time embeds when rescaling timesteps, because using the corrected CFG rescale code only marginally improves the quality of the image:

Diffusers main:

![image](https://github.com/huggingface/diffusers/assets/59658056/feea3789-bfda-4dd5-a754-9197f509556d)

With change:

![image](https://github.com/huggingface/diffusers/assets/59658056/145c3edc-9f01-4fc6-892b-af70b68a2645)

**Comment by sayakpaul** [comment]
Maybe @patil-suraj has something to offer. He's been running a couple of experiments that involve a flavor of how SDXL was originally trained. 

The pix2pix SDXL script has a bug which I found out yesterday. Will open a PR soon :-)

**Comment by sayakpaul** [comment]
Nope. Separate one.

**Comment by bghira** [comment]
this is SDXL in general, as i've seen it on the pix2pix script as well as text2img, and even other projects have reported the same "milky" image quality issue.

for me the issue is real as I can definitely fine-tune SD 2.1 just fine on this noise schedule.

if the issue can be reproduced on, you know, batch size 1024 or some scale of operations i am incapable of testing with right now, the answer would then be that i am simply undersizing my hw and i will accept that answer.

but it just feels odd, even with low learning rate and long time, that it never adapts

**Comment by williamberman** [comment]
Sorry can you elaborate on the "I can definitely fine-tune SD 2.1 just fine on this noise schedule"? You're saying you have no issues finetuning without v-prediction?

**Comment by bghira** [comment]
i've adapted 2.0-base to v_prediction as an experiment, with a batch size of 150 and an adaptive cosine with restarts learning rate from 4e-8 to 4e-7.

i've fine-tuned 2.0-v, and i've fine-tuned 2.1-v.

but i've not adapted 2.1-base to v_prediction.

**Comment by bghira** [comment]
my fine-tuning has been done with the scripts provided by the examples directory as well as [this code](https://github.com/bghira/SimpleTuner/tree/release) that I maintain, which is based on the example scripts from the repo here and makes use of Diffusers unet and scheduler code.

it's the same tests that work on 2.x and do not on SDXL, which makes me think maybe we/I have handled the added_cond_kwargs incorrectly. I really don't know how the scaled timestep schedule interacts with the time embeds.

additionally, I understand that the noise curve of 'trailing' timestep config with and without rescaling betas for zero SNR match up until the final ~700-1000 timestep window, where the noise schedule really needs to apply. it's possible I'm just not hitting those. 

My training logs are [on Weights and Biases](https://wandb.ai/bghira/projects)

**Comment by williamberman** [comment]
Ah ok so the tl;dr is finetuning using v-prediction on the non xl variants works but have had issues just with xl?

**Comment by bghira** [comment]
yep. i've gotten close twice, without using dropout probability:

on 8k kodachrome images
![image](https://github.com/huggingface/diffusers/assets/59658056/f1f1c8a4-87d5-4abc-9572-7329298aaf8c)


and 140k COCO photos
![image](https://github.com/huggingface/diffusers/assets/59658056/a1e06104-6063-49ae-8eed-95203e127a98)

there still seems to be some ultimate floor of contrast where darks stop improving, and then the model diverges.

**Comment by bghira** [comment]
Linear histogram

![image](https://github.com/huggingface/diffusers/assets/59658056/dc2018df-fb70-4fc3-9b15-05e8d6a91b33)

**Comment by PeterL1n** [comment]
It looks like to me that when you train your model, you normalized the images to 0~1 range instead of -1~1 range. So the model is trained to produce 0~1 range, but at inference you treat it as -1~1 range for rescale, so the images now has 0.5~1 range.

**Comment by bghira** [comment]
i see now. it looks like this wasn't getting called correctly :+1: I will try again with new latents.

**Comment by bghira** [comment]
Haven't verified yet

**Comment by bghira** [comment]
![image](https://github.com/huggingface/diffusers/assets/59658056/2c9f8ac8-e440-4956-8364-ae1f7863f921)

Evidently bypassing the image cropping transforms meant I lost the nornalization too. this one is 100% on me, i caused this problem. thank you all for the help.

**Comment by sayakpaul** [comment]
Fire! 

When you feel good, please do share the fine-tuned checkpoints too! Would love to ⏯️.

**Comment by bghira** [comment]
the current plan is to finish up a proof-of-concept checkpoint here, and then make a few follow-up fixes to the dataset sampler code, before settling on a longer training run, with a greater quantity of images.

SDXL trains impressively fast. EMA weights and min-SNR are a godsend. you don't get v-prediction adaptation this quickly with AdamW8Bit without min-SNR.

375 steps into training:

![image](https://github.com/huggingface/diffusers/assets/59658056/f7aae692-bc5b-4cd5-9b3c-748499b258c5)

Our WandB logs (ignore the incorrect tag in the project name, this uses AdamW8Bit, not Dadapt)
https://wandb.ai/bghira/sdxl-v-cocoPhoto87k-minSNR-dadapt-norm

**Comment by sayakpaul** [comment]
The code is still from your SimpleTuner repo?

**Comment by bghira** [comment]
yes :)

**Comment by bghira** [comment]
I actually totally rewrote the bucketing logic this weekend, so that others as yourself can read it more easily with less time invested in following the breadcrumbs.

**Comment by sayakpaul** [comment]
I would probably be very interested in a simple notebook taking your bucketing-enabled loader to visualize and investigate! Thank you for your contributions to the community!

**Comment by bghira** [comment]
I need to sit down with someone at some point and figure out what is needed to make this thing multi-GPU capable for the dataloader. I know that your upstream examples tend to select an iterator based on whether data parallel should be used, or not. but the documentation on that is kind of scarce. do you have a good resource I can look at?

**Comment by sayakpaul** [comment]
> I need to sit down with someone at some point and figure out what is needed to make this thing multi-GPU capable for the dataloader.

Oh okay. So, is the current run going on a single GPU? I think that is okay for a PoC. 

For the reference, I would recommend the following the dataloader approach presented in:
https://arxiv.org/abs/2110.02178

**Comment by bghira** [comment]
> Oh okay. So, is the current run going on a single GPU? I think that is okay for a PoC.

yes, there's some issues that will take a while to work out.

while i took a look at the referenced code repo, it looks like they require heavy pre-processing of the datasets, and my dataloader tries to make it a non-issue. I simply provide nested folders of images.
